### DISCUSSION TITLE
Script : Create hyperlink to OS command

### DISCUSSION THREAD

**Discussion by lilive** [freeplane/freeplane#1039] [Help] [answered]
Hello,

In a script, I want to create a hyperlink to execution of an OS command, like the menu entry `Insert > Link > Add hyperlink to execution of an OS command...` do.

I do not see which method I could use in the API.

I know that it is possible to do something like
```groovy
node.link.text = "execute:_notepad"
```
but sometimes I have problems because some characters need special encoding.

For example, `Menu > Insert > Link > Add hyperlink to execution of an OS command...` translate the command `"C:\Program Files\7-Zip\7zFM.exe"` to the link `execute:_%22C:%5CProgram%20Files%5C7-Zip%5C7zFM.exe%22`.

Does anyone know how we can have the same result with a script ?

Thank you.

**Comment by lilive** [comment]
PS:

In the past I tried to write an encoding function like that, but it may fail, for example if they are accentuated letters in the command line:
```groovy
// Replace some characters with their percent-encoded value
String encodeCommand( String cmd ){
    // This is experimental, I do not know exactly which characters need to be replaced,
    // I'm just trying to mimic what I've seen when using the built-in menu entry
    // Insert > Link > OS Command
    def check = ~/[\d\w-_\.:]/
    String output = ""
    for( c in cmd ){
        if( c =~ check ) output += c
        else output += "%" + Integer.toHexString( (int)c.charAt(0) ).toUpperCase()
    }
    return output
}

// Problem with accents !
node.link.text = encodeCommand( /execute:_"C:\Program Files\SumatraPDF\SumatraPDF.exe" "C:\docs\Stéréovision axiale.pdf"/ )
```

In this example the link is created with the string
`execute:_%22C:%5CProgram%20Files%5CSumatraPDF%5CSumatraPDF.exe%22%20%22C:%5Cdocs%5CSt%E9r%E9ovision%20axiale.pdf%22`
and when I follow it, SumatraPDF starts and complains it can open `C:\docs\St�r�ovision axiale.pdf`

I guess I can find a way to adapt my `encodeCommand` function to not touch the accentuated letters, but I'm looking for a reliable solution that will work in all cases.

**Comment by EdoFro** [comment]
I think, but I'm not sure, that htmlUtils has some methods that may help you.

**Reply by lilive** [reply]
I have tried all the methods that seem relevant to me, with no luck.

**Comment by EdoFro** [comment, marked as answer]
If not, you can create a new File  object with the path and then get its Uri and modify the uri components to get what you want.

Just an idea.

**Reply by lilive** [reply]
File is not the perfect fit, because I want to start programs with command line arguments. But URI seems to be a solution !

```groovy
import java.net.URI
cmd = /"C:\Program Files\SumatraPDF\SumatraPDF.exe" "C:\docs\Stéréovision axiale.pdf"/
node.link.uri = new URI( 'execute', "_" + cmd, null )
```

This seems a little bit odd (I don't understand why I have to prepend the underscore like that) but it give me an usable hyperlink command. Thank you Edo.

**Comment by EdoFro** [comment]
> or example, `Menu > Insert > Link > Add hyperlink to execution of an OS command...` translate the command `"C:\Program Files\7-Zip\7zFM.exe"` to the link `execute:_%22C:%5CProgram%20Files%5C7-Zip%5C7zFM.exe%22`.

also. the link text is uri, so you should use normal slashes "/" in the link's path

**Comment by abc16361** [comment]
If I need to add a hyperlink to a node which can activate an application when clicked, I type this: 
`node.link.text = "file:/C:/Program%20Files/7-Zip/7zFM.exe"`
If I need to execute the app within the script, I use:
`"\"C:/Program Files/7-Zip/7zFM.exe\" ".execute()`
Or use this to open a file:
`"\"C:/Program Files/Inkscape/bin/inkscape.exe\" file:///F:/drawTest.svg".execute()`

**Reply by lilive** [reply]
You helped me to think, thank you.

> If I need to add a hyperlink to a node which can activate an application when clicked, I type this:
`node.link.text = "file:/C:/Program%20Files/7-Zip/7zFM.exe"`

Good to know. Executable are files, indeed, and I never thought that I could use `file:` links like that.

> use this to open a file:
`"\"C:/Program Files/Inkscape/bin/inkscape.exe\" file:///F:/drawTest.svg".execute()`

This is really convenient. Nevertheless my goal is to create all sort of links that point to various resources. For example to open a pdf with my pdf reader, at a specific page. I like the hyperlinks to OS command feature, because these links can be visited like the other type of hyperlinks (websites, files), by pressing Ctrl+Enter.
